### PR TITLE
fix string parsing

### PR DESCRIPTION
**Pull request by ZacLN**

### CODE REVIEW

**Review by davidanthoff — DISMISSED** [review]

File: src/lexer.jl [hunk 1]
```diff
@@ -40,10 +40,18 @@ mutable struct ParseState
...
 function ParseState(str::Union{IO,String})
-    ps = ParseState(tokenize(str, RawToken), false, RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), Closer(), false)
+    lastbyte = if str isa String
+        sizeof(str)
+    elseif str isa IOBuffer
+        str.size
+    else
+        Inf
+    end
+    ps = ParseState(tokenize(str, RawToken), false, RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), Closer(), str isa String ? sizeof(str) : lastbyte, false)
```
Comment (pfitzseb) [inline src/lexer.jl:54]:
```suggestion
    ps = ParseState(tokenize(str, RawToken), false, RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), Closer(), lastbyte, false)
```

File: src/lexer.jl [hunk 2]
```diff
@@ -40,10 +40,18 @@ mutable struct ParseState
...
     errored::Bool
 end
 function ParseState(str::Union{IO,String})
-    ps = ParseState(tokenize(str, RawToken), false, RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), RawToken(), Closer(), false)
+    lastbyte = if str isa String
+        sizeof(str)
+    elseif str isa IOBuffer
+        str.size
+    else
+        Inf
```
Comment (pfitzseb) [inline src/lexer.jl:52]: Probably best to error here, no?

Comment (ZacLN) [reply src/lexer.jl:52]: I think I'll just restrict the signature (same effect)

**Comment by ZacLN** [comment]
@davidanthoff, if you intend to push out a new insiders version in the next few days we should merge this just beforehand. The tests are still failing (and I'll continue to try to squash them) but they're very much corner case, and it would be good to get the fixes as they are out into the wild.

**Comment by davidanthoff** [comment]
Lets say we push the next insider build once this PR passes test and is merged then.

**Comment by ZacLN** [comment]
Well that was a slog. This is ready

File: src/components/strings.jl [hunk 3]
```diff
@@ -149,23 +153,43 @@ function parse_string_or_cmd(ps::ParseState, prefixed=false)
...
+                            pushtotrivia!(ret, rparen)
+                            seek(input, ps1.nt.startbyte + 1)
+                        else
+                            pushtotrivia!(ret, EXPR(:RPAREN, 0, 0))
+                            if eof(input)
+                                # Some errored interpolation, we're now at the end of the string but lets make sure we have the closing "
+                                # write(b, '"')
+                                # erroredonlast = true
+                                # @info "break"
+                            #    break
```
Comment (davidanthoff) [inline src/components/strings.jl:171]: Is this meant to be carried forward, or could this be deleted? If it should stay, maybe add a bit more explanation what it is for some future person trying to understand this :)

**Review by davidanthoff — APPROVED** [review]
One small request that you should feel free to ignore ;)

File: src/components/strings.jl (hunk 3 above)
Comment (ZacLN) [reply src/components/strings.jl]: Nope just some leftover junk!

**Comment by davidanthoff** [comment]
Hm, now tests fail again? Weird...

**Review by davidanthoff — APPROVED** [review]

**Comment by ZacLN** [comment]
The tests now check for correctness (of various description)  when mutating every file in the package 1000 times (e.g. deleteing  tokens), so there's a bit of randomness about the place.

**Comment by davidanthoff** [comment]
Maybe we should set a seed then at the beginning of the random part?
